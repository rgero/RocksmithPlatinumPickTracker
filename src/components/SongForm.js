import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import ErrorModal from './ErrorModal';

export default class SongForm extends React.Component {

    constructor(props)
    {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onCalendarFocusChanged = this.onCalendarFocusChanged.bind(this);
        this.onAccuracyChange = this.onAccuracyChange.bind(this);
        this.onPickLevelChange = this.onPickLevelChange.bind(this);
        this.clearError = this.clearError.bind(this);

        this.state = {
            id: this.props.songInfo ? this.props.songInfo.id : "",
            artist: this.props.songInfo ? this.props.songInfo.artist : "",
            songName: this.props.songInfo ? this.props.songInfo.songName : "",
            date: this.props.songInfo ? moment(this.props.songInfo.date) : moment(),
            path: this.props.songInfo ? this.props.songInfo.path : "",
            difficulty: this.props.songInfo ? this.props.songInfo.difficulty : "",
            level: this.props.songInfo ? this.props.songInfo.level : "platinum",
            accuracy: this.props.songInfo ? (this.props.songInfo.accuracy/100).toString() : "",
            notes: this.props.songInfo ? this.props.songInfo.notes : "",
            calendarFocused: false,
            error: "",
        }

    }

    onSubmit(e){
        e.preventDefault();

        //Validate Form;
        var currentError = "";
        if (!this.state.artist || !this.state.songName || !this.state.date || !this.state.path || !this.state.accuracy || !this.state.difficulty)
        {
            currentError = "Please inspect the form";
        }
        
        if( currentError === ""){
            this.setState({error: currentError})
            this.props.onSubmit({
                id: this.state.id,
                artist: this.state.artist,
                songName: this.state.songName,
                date: this.state.date.valueOf(),
                path: this.state.path,
                difficulty: this.state.difficulty,
                level: this.state.level,
                accuracy: this.state.accuracy,
                notes: this.state.notes
            });
        } else {
            this.setState({error: currentError})
        }
    }

    onTextChange = (value) => (evt) => {
        var newValue = evt.target.value;
        switch(value){
            case "artist":
                this.setState({
                    artist: newValue
                });
                break;
            case "songName":
                this.setState({
                    songName: newValue
                });
                break;
            case "path":
                this.setState({
                    path: newValue
                });
                break;
            case "difficulty":
                this.setState({
                    difficulty: newValue
                });
                break;
            case "notes":
                this.setState({
                    notes: newValue
                });
                break;
            default:
                break;
        }
    }

    onDateChange(newDate){
        if(newDate){
            this.setState(() => ({date: newDate}));
        }
    }

    onCalendarFocusChanged(calendarFocused){
        calendarFocused = calendarFocused["focused"]
        this.setState(() => ({calendarFocused}));
    }

    onAccuracyChange(e){
        const amount = e.target.value;
        if ((!amount || amount.match(/^\d{1,3}(\.\d{0,2})?$/)) && (!amount || parseFloat(amount) <= 100)) {
            this.setState(()=>({
                accuracy: amount
            }));
        }
    }

    onPickLevelChange(e){
        const value = e.target.value
        this.setState(()=>({
            level: value
        }))
    }

    clearError(){
        this.setState({error: ""})
    }

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error !== '' && <ErrorModal errorMessage={this.state.error} clearError={this.clearError} />}
                
                {/* Artist */}
                <div className = "form__input">
                    <label className = "form__label">Artist</label>
                    <input
                        type="text"
                        className="text-input"
                        placeholder="Artist"
                        autoFocus
                        value={this.state.artist}
                        onChange={this.onTextChange("artist")}
                    />
                </div>

                {/* Song Name */}
                <div className = "form__input">
                    <label className = "form__label">Song Name</label>
                    <input
                        type="text"
                        className="text-input"
                        placeholder="Song Name"
                        autoFocus
                        value={this.state.songName}
                        onChange={this.onTextChange("songName")}
                    />
                </div>

                {/* Date */}
                <div className = "form__input">
                    <label className = "form__label">Date</label>
                    <SingleDatePicker 
                        date={this.state.date}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChanged}
                        numberOfMonths={1}
                        isOutsideRange={()=> false}
                        block
                    />
                </div>

                {/* Path */}
                <div className = "form__input">
                    <label className = "form__label">Path</label>
                    <input
                        type="text"
                        className="text-input"
                        placeholder="Path"
                        autoFocus
                        value={this.state.path}
                        onChange={this.onTextChange("path")}
                    />
                </div>

                {/* Pick Level */}
                <div className="form__input">
                    <label className = "form__label">Pick Level</label>
                    <select value={this.state.level}
                            onChange={this.onPickLevelChange}
                    >
                        <option value="platinum">Platinum</option>
                        <option value="gold">Gold</option>
                        <option value="silver">Silver</option>
                        <option value="bronze">Bronze</option>
                    </select>
                </div>

                {/* Difficulty */}
                <div className = "form__input">
                    <label className = "form__label">Difficulty</label>
                    <input
                        type="text"
                        className="text-input"
                        placeholder="Difficulty"
                        autoFocus
                        value={this.state.difficulty}
                        onChange={this.onTextChange("difficulty")}
                    />
                </div>

                {/* Accuracy */}
                <div className = "form__input">
                    <label className = "form__label">Accuracy</label>
                    <input
                        type="text"
                        className="text-input"
                        placeholder="Accuracy"
                        autoFocus
                        value={this.state.accuracy}
                        onChange={this.onAccuracyChange}
                    />
                </div>

                {/* Notes */}
                <div className = "form__input">
                    <label className = "form__label">Notes</label>
                    <textarea 
                        placeholder="Notes about the song (optional)"
                        className="textarea"
                        value={this.state.note}
                        onChange={this.onTextChange("notes")}
                    >
                    </textarea>
                </div>

                <div className = "form__input">
                    <button className="button">Save Entry</button>
                </div>
            </form>
        )
    }
}