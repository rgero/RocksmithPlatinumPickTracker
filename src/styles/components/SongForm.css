.form {
    display: flex;
    flex-direction: column;
    > * {
        margin-bottom: 1.6rem;
    }
}

.form__error {
    margin: 0 0 1.6rem 0;
    font-style: italic;
}
  
.form__input{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    padding-bottom: 1.6rem;
    align-items: center;
    width: 50%;
}

.form__label {
    font-weight:bold;
    margin-bottom: 0;
}

.textarea {
    min-width: 75%;
    min-height: 1.25em;
    line-height: 1.25;
}