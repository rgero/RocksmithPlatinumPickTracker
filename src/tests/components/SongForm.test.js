import React from 'react';
import {shallow} from 'enzyme';
import moment from 'moment';

import SongForm from '../../components/SongForm';

test("Rendering the Song Form Form", ()=>{
    const wrapper =shallow(<SongForm/>);
    expect(wrapper).toMatchSnapshot();
})

// Testing the Inputs
test("Input - Artist change", ()=>{
    const value = "Rise Against"
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(0) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('artist')).toBe(value);
})

test("Input - Song change", ()=>{
    const value = "Far from Perfect"
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(1) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('songName')).toBe(value);
})

test("Testing Date change", ()=> {
    const now = moment();
    const wrapper = shallow(<SongForm />);
    const testObj = wrapper.find('withStyles(SingleDatePicker)');
    expect(testObj.length).toBe(1);
    testObj.prop("onDateChange")(now);
    expect(wrapper.state('date')).toEqual(now);
})


test("Input - Path change", ()=>{
    const value = "Alternate Lead"
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(2) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('path')).toBe(value);
})

test("Input - Difficulty change", ()=>{
    const value = "Hard"
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(3) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('difficulty')).toBe(value);
})

test("Input - Note change", ()=>{
    const value = "This song was really fun, I should play it more";
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('textarea').at(0) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('notes')).toBe(value);
})

test("Input - Accuracy Change - Valid Input", ()=>{
    const value = "91.23";
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(4) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('accuracy')).toBe(value);
})

test("Input - Accuracy Change - Invalid Input (Not Number)", ()=>{
    const value = "Alpha";
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(4) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('accuracy')).toBe(""); // Empty String is the current default value
})

test("Input - Accuracy Change - Invalid Input (Too long)", ()=>{
    const value = "10213.12";
    const wrapper = shallow(<SongForm />);
    const descriptionElem = wrapper.find('input').at(4) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('accuracy')).toBe(""); // Empty String is the current default value
})

test("Input - Pick Level Change", ()=>{
    const value = "Bronze";
    const wrapper = shallow(<SongForm />);
    expect(wrapper.state('level')).toBe("Platinum");
    const descriptionElem = wrapper.find('select').at(0) 
    descriptionElem.simulate('change', {
        target: {value}
    })
    expect(wrapper.state('level')).toBe(value); // Empty String is the current default value
})