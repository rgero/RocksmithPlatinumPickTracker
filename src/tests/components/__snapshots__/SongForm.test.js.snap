// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the Song Form Form 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Artist
    </label>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Artist"
      type="text"
      value=""
    />
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Song Name
    </label>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Song Name"
      type="text"
      value=""
    />
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Date
    </label>
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={true}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Path
    </label>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Path"
      type="text"
      value=""
    />
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Pick Level
    </label>
    <select
      onChange={[Function]}
      value="Platinum"
    >
      <option
        value="Platinum"
      >
        Platinum
      </option>
      <option
        value="Gold"
      >
        Gold
      </option>
      <option
        value="Silver"
      >
        Silver
      </option>
      <option
        value="Bronze"
      >
        Bronze
      </option>
    </select>
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Difficulty
    </label>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Difficulty"
      type="text"
      value=""
    />
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Accuracy
    </label>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Accuracy"
      type="text"
      value=""
    />
  </div>
  <div
    className="form__input"
  >
    <label
      className="form__label"
    >
      Notes
    </label>
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Notes about the song (optional)"
      value=""
    />
  </div>
  <div
    className="form__input"
  >
    <button
      className="button"
    >
      Save Entry
    </button>
  </div>
</form>
`;
